codebleu/codebleu_metric.py

```diff
@@ -0,0 +1,308 @@
+
+
+def _find_node(G, id):  # to find a node in the graph by the corresponding variable id (same to location)
+    for node in list(G.nodes):
+        if node not in G.nodes:
+            return None
+        if G.nodes[node]['id'] == id:
+            return node
+    return None
+
+
+def _find_edge(G, edge_data):  # find the edge from the candidate graph in the reference graph
+    edge_list = list(G.edges)
+    edge_list.sort(key=lambda x: x[0])
+    for edge in edge_list:
+        if (G.nodes[edge[0]]['data'] == edge_data[0]) and (G.nodes[edge[1]]['data'] == edge_data[1]):
+            return edge
+    return None
+
+
+def _find_matching_variable(variables_dict,
+                            thekey):
+    # if there are more than one variables with the same name in the code
+    # (e.g. we have several functions with same names of local variables),
+    # we wanna parse, to which variable do we refer now
+    lookup = _split_identifier(thekey)
+    match_var = [0, '']
+    for key, _ in variables_dict.items():
+        checked = _split_identifier(key)
+        if (lookup[1] == checked[1]) and (lookup[0] >= checked[0]) and (match_var[0] < checked[0]):
+            match_var = checked
+    return [lookup[0], str(match_var[0]) + '_' + match_var[1]]
+
+
+def _create_graph(next_use_dict, variables_dict, occur_dict):
+    G = nx.DiGraph()
+    nodes_list = []
+    for key, value in occur_dict.items():
+        toadd = _find_matching_variable(variables_dict, key)
+        if toadd[1] != '0_':  # to disregard the attributes -- afaiu, we don't need them
+            nodes_list.append(toadd)
+    nodes_list.sort(key=lambda x: x[0])
+    for i, item in enumerate(nodes_list):
+        G.add_node(i, id=item[0], data=variables_dict[item[1]])
+    for key, value in next_use_dict.items():
+        from_node = _find_node(G, _split_identifier(key)[0])
+        for item in value:
+            to_node = _find_node(G, _split_identifier(item)[0])
+            G.add_edge(from_node, to_node)
+    return G
+
+
+def _compare_graphs(ref_G,
+                    cand_G):  # we look up edges from candidate one by one, if they are in the reference; if they are
+    # -- increase # of matches by one and pop the corresponding edge from list of ref edges
+    matched = 0
+    cand_edges = list(cand_G.edges)
+    cand_edges.sort(key=lambda x: x[0])
+    for edge in cand_edges:
+        edge_data = (cand_G.nodes[edge[0]]['data'], cand_G.nodes[edge[1]]['data'])
+        matched_edge = _find_edge(ref_G, edge_data)
+        if matched_edge is not None:
+            ref_G.remove_edge(*matched_edge)
+            matched += 1
+    return matched
+
+
+def dfg_match(reference, candidate, lattice):  # dfg part of the metric
+    try:
+        ref_parsed = _fix_graph(gg.AstGraphGenerator(reference, lattice).build())
+    except:
+        return -1  # we return -1 if it is impossible to do some step for reference snippet and return 0 if it is possible for reference, but not for candidate
```

Comment:
Let's create an `Enum` in order to improve readability ([see here](https://docs.python.org/3/library/enum.html)) or raise an error and catch it afterward. Handling constants in such cases is a bit messy.